topics/reference_guide/custom_language_support/language_server_protocol.md

```diff
@@ -0,0 +1,183 @@
+- Code completion ([textDocument/completion](https://microsoft.github.io/language-server-protocol/specification/#textDocument_completion))
+- Go to Declaration ([textDocument/definition](https://microsoft.github.io/language-server-protocol/specification/#textDocument_definition))
+- More features will be announced soon.
+
+## Basic implementation
+
+A minimal LSP integration must implement `LspServerSupportProvider` along with a service descriptor and register it as a `com.intellij.platform.lsp.serverSupportProvider` extension point:
+
+```kotlin
+import com.intellij.platform.lsp.api.LspServerSupportProvider
+import com.intellij.platform.lsp.api.ProjectWideLspServerDescriptor
+
+class FooLspServerSupportProvider : LspServerSupportProvider {
+  override fun fileOpened(project: Project, file: VirtualFile, serverStarter: LspServerStarter) {
+    if (file.extension == "foo") {
+      serverStarter.ensureServerStarted(FooLspServerDescriptor(project))
+    }
+  }
+}
+
+private class FooLspServerDescriptor(project: Project) : ProjectWideLspServerDescriptor(project, "Foo") {
+  override fun isSupportedFile(file: VirtualFile) = file.extension == "foo"
+  override fun createCommandLine() = GeneralCommandLine("foo", "--stdio")
+}
+```
+
+After providing the implementation of the LSP Server support provider extension point, register it in the <path>plugin.xml</path> file as follows:
+
+```xml
+<idea-plugin>
+   <!-- ... -->
+   <depends>com.intellij.modules.platform</depends>
+   <depends>com.intellij.modules.ultimate</depends>
+   <extensions defaultExtensionNs="com.intellij">
+       <platform.lsp.serverSupportProvider implementation="FooLspServerSupportProvider"/>
+   </extensions>
+</idea-plugin>
+```
+
+As a reference, check out the [Prisma ORM](https://plugins.jetbrains.com/plugin/20686-prisma-orm) open-source plugin implementation: https://github.com/JetBrains/intellij-plugins/tree/master/prisma/src/org/intellij/prisma/ide/lsp
+
+## Language Server integration
+
+Language Server is a separate process that analyzes source code and provides language-specific features to development tools.
+When creating a plugin that utilizes LSP within the IDE, there are two possibilities for providing a Language Server to end-users:
+
+- Bundle a [Language Server implementation](https://microsoft.github.io/language-server-protocol/implementors/servers/) binary as a resource delivered with a plugin.
+- Provide a possibility for users to define the location of the Language Server binary in their environment.
+
+The Prisma ORM plugin presents the first approach, which distributes the `prisma-language-server.js` script and uses a local Node.js interpreter to run it.
+
+For more complex cases, you may request to provide a detailed configuration with a dedicated [Settings](settings_guide.md) implementation.
+
+To implement a minimal but fully functional plugin, perform the following steps:
```

Comment:
I think we don't need this paragraph at all. It is already described in the _Basic Implementation_ section. It doesn't add anything new here. Maybe we should just be more descriptive in the previous section about the steps, but IMHO there is no need, as there are excellent Javadocs.